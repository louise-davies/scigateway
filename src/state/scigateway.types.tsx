import { Step } from 'react-joyride';
import { Theme } from '@material-ui/core/styles/createMuiTheme';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadedAuthType = 'scigateway:loaded_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadScheduledMaintenanceStateType =
  'scigateway:load_scheduled_maintenance_state';
export const LoadMaintenanceStateType = 'scigateway:load_maintenance_state';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SendThemeOptionsType = 'scigateway:api:send_themeoptions';
export const LoadDarkModePreferenceType =
  'scigateway:load_dark_mode_preference';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';
export const RegisterStartUrlType = 'scigateway:register_start_url';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
  showHelpPageButton: boolean;
}

export interface StartUrlPayload {
  startUrl: string;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface SendThemeOptionsPayload {
  theme: Theme;
  broadcast: boolean;
}

export interface LoadDarkModePreferencePayload {
  darkMode: boolean;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface ScheduledMaintenanceStatePayLoad {
  scheduledMaintenance: ScheduledMaintenanceState;
}

export interface ScheduledMaintenanceState {
  show: boolean;
  message: string;
}

export interface MaintenanceStatePayLoad {
  maintenance: MaintenanceState;
}

export interface MaintenanceState {
  show: boolean;
  message: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
