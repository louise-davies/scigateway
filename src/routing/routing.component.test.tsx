import React from 'react';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { createLocation } from 'history';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import * as singleSpa from 'single-spa';
import { ThemeProvider } from '@mui/material';
import Routing, { PluginPlaceHolder } from './routing.component';
import TestAuthProvider from '../authentication/testAuthProvider';
import { StateType } from '../state/state.types';
import { authState, initialState } from '../state/reducers/scigateway.reducer';
import { buildTheme } from '../theming';

jest.mock('../adminPage/adminPage.component', () => () => 'Mocked AdminPage');
jest.mock(
  '../maintenancePage/maintenancePage.component',
  () => () => 'Mocked MaintenancePage'
);
jest.mock('../preloader/preloader.component', () => ({
  Preloader: () => 'Mocked Preloader',
}));
jest.mock('single-spa', () => ({
  unloadApplication: jest.fn(),
}));

describe('Routing component', () => {
  let mockStore;
  let state: StateType;
  const theme = buildTheme(false);

  beforeEach(() => {
    state = {
      scigateway: { ...initialState, authorisation: { ...authState } },
      router: {
        action: 'POP',
        location: createLocation('/'),
      },
    };

    mockStore = configureStore();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('renders component with no plugin routes', () => {
    state.scigateway.plugins = [];
    const wrapper = shallow(<Routing store={mockStore(state)} />)
      .dive()
      .dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders component with plugins', () => {
    state.scigateway.plugins = [
      {
        section: 'test section',
        link: 'test link',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin',
        order: 1,
      },
      {
        section: 'test section',
        link: 'test link alt',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin Alt link',
        order: 2,
      },
      {
        section: 'test section 2',
        link: 'test link 2',
        plugin: 'test_plugin_name_2',
        displayName: 'Test Plugin 2',
        order: 3,
      },
      {
        section: 'test section',
        link: 'test link admin',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin Admin',
        admin: true,
        order: 4,
      },
      {
        section: 'test section',
        link: 'test link admin alt',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin Admin Alt link',
        admin: true,
        order: 5,
      },
    ];
    const wrapper = shallow(<Routing store={mockStore(state)} />)
      .dive()
      .dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a route for a plugin when site is under maintenance and user is admin', () => {
    state.scigateway.authorisation.provider = new TestAuthProvider('logged in');
    state.scigateway.siteLoading = false;
    state.scigateway.maintenance = { show: true, message: 'test' };
    state.scigateway.plugins = [
      {
        section: 'test section',
        link: '/test_link',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin',
        order: 1,
      },
    ];
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter
          initialEntries={[{ key: 'testKey', pathname: '/test_link' }]}
        >
          <ThemeProvider theme={theme}>
            <Routing />
          </ThemeProvider>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a route for maintenance page when site is under maintenance and user is not admin', () => {
    const testAuthProvider = new TestAuthProvider('logged in');
    testAuthProvider.isAdmin = jest.fn().mockImplementationOnce(() => false);
    state.scigateway.authorisation.provider = testAuthProvider;
    state.scigateway.siteLoading = false;
    state.scigateway.maintenance = { show: true, message: 'test' };
    state.scigateway.plugins = [
      {
        section: 'test section',
        link: '/test_link',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin',
        order: 1,
      },
    ];
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter
          initialEntries={[{ key: 'testKey', pathname: '/test_link' }]}
        >
          <ThemeProvider theme={theme}>
            <Routing />
          </ThemeProvider>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders placeholder for a plugin', () => {
    const wrapper = shallow(<PluginPlaceHolder id="test_id" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a route for admin page', () => {
    state.scigateway.authorisation.provider = new TestAuthProvider('logged in');
    state.scigateway.siteLoading = false;
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter initialEntries={[{ key: 'testKey', pathname: '/admin' }]}>
          <ThemeProvider theme={theme}>
            <Routing />
          </ThemeProvider>
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('redirects to a homepage URL if specified', () => {
    state.scigateway.homepageUrl = '/homepage';

    const wrapper = shallow(<Routing store={mockStore(state)} />)
      .dive()
      .dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('redirects to the homepage if navigating to login page while logged in', () => {
    state.scigateway.authorisation.provider.isLoggedIn = jest
      .fn()
      .mockImplementationOnce(() => true);

    state.scigateway.authorisation.provider.autoLogin = jest
      .fn()
      .mockImplementationOnce(() => Promise.reject());

    window.localStorage.__proto__.getItem = jest
      .fn()
      .mockImplementationOnce((name) =>
        name === 'autoLogin' ? 'false' : null
      );

    const wrapper = shallow(<Routing store={mockStore(state)} />)
      .dive()
      .dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('single-spa remounts a plugin when switching between admin and non-admin plugins via single-spa:before-no-app-change event', () => {
    state.scigateway.authorisation.provider = new TestAuthProvider('logged in');
    state.scigateway.siteLoading = false;
    state.scigateway.plugins = [
      {
        section: 'test section',
        link: '/test_link',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin',
        order: 1,
      },
      {
        section: 'test section',
        link: '/admin_test_link',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin Admin',
        admin: true,
        order: 2,
      },
    ];
    mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter initialEntries={['/test_link']}>
          <Routing />
        </MemoryRouter>
      </Provider>
    );

    window.dispatchEvent(
      new CustomEvent('single-spa:before-no-app-change', {
        detail: {
          oldUrl: 'http://localhost/test_link',
          newUrl: 'http://localhost/admin_test_link',
        },
      })
    );
    expect(singleSpa.unloadApplication).toHaveBeenCalledWith(
      'test_plugin_name'
    );

    (singleSpa.unloadApplication as jest.Mock).mockClear();

    window.dispatchEvent(
      new CustomEvent('single-spa:before-no-app-change', {
        detail: {
          oldUrl: 'http://localhost/admin_test_link',
          newUrl: 'http://localhost/test_link',
        },
      })
    );
    expect(singleSpa.unloadApplication).toHaveBeenCalledWith(
      'test_plugin_name'
    );
  });

  it("single-spa reloads a plugin when it hasn't loaded for some reason", () => {
    jest.useFakeTimers();
    state.scigateway.authorisation.provider = new TestAuthProvider('logged in');
    state.scigateway.siteLoading = false;
    state.scigateway.plugins = [
      {
        section: 'test section',
        link: '/test_link',
        plugin: 'test_plugin_name',
        displayName: 'Test Plugin',
        order: 1,
      },
    ];
    state.router.location = createLocation('/test_link');

    jest.spyOn(document, 'getElementById').mockImplementation(() => {
      return document.createElement('div');
    });

    const clearIntervalSpy = jest.spyOn(window, 'clearInterval');

    mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter initialEntries={['/test_link']}>
          <Routing />
        </MemoryRouter>
      </Provider>
    );
    jest.runAllTimers();

    expect(singleSpa.unloadApplication).toHaveBeenCalledWith(
      'test_plugin_name'
    );

    expect(clearIntervalSpy).toHaveBeenCalledWith(expect.any(Number));
  });
});
