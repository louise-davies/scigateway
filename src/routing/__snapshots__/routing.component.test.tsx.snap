// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routing component redirects to a homepage URL if specified 1`] = `
<ContextProvider
  value={null}
>
  <Routing
    dispatch={[Function]}
    drawerOpen={false}
    homepageUrl="/homepage"
    loading={true}
    location="/"
    maintenance={
      Object {
        "message": "",
        "show": false,
      }
    }
    plugins={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    userIsAdmin={false}
    userIsloggedIn={false}
  />
</ContextProvider>
`;

exports[`Routing component redirects to the homepage if navigating to login page while logged in 1`] = `
<ContextProvider
  value={null}
>
  <Routing
    dispatch={[Function]}
    drawerOpen={false}
    loading={true}
    location="/"
    maintenance={
      Object {
        "message": "",
        "show": false,
      }
    }
    plugins={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    userIsAdmin={false}
    userIsloggedIn={true}
  />
</ContextProvider>
`;

exports[`Routing component renders a route for a plugin when site is under maintenance and user is admin 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "testKey",
          "pathname": "/test_link",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/test_link",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/test_link",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Routing)>
        <Routing
          dispatch={[Function]}
          drawerOpen={false}
          loading={false}
          location="/"
          maintenance={
            Object {
              "message": "test",
              "show": true,
            }
          }
          plugins={
            Array [
              Object {
                "displayName": "Test Plugin",
                "link": "/test_link",
                "order": 1,
                "plugin": "test_plugin_name",
                "section": "test section",
              },
            ]
          }
          userIsAdmin={true}
          userIsloggedIn={true}
        >
          <Styled(div)
            drawerOpen={false}
          >
            <Insertion
              cache={
                Object {
                  "insert": [Function],
                  "inserted": Object {
                    "jlgjno": true,
                  },
                  "key": "css",
                  "nonce": undefined,
                  "registered": Object {},
                  "sheet": StyleSheet {
                    "_alreadyInsertedOrderInsensitiveRule": true,
                    "_insertTag": [Function],
                    "before": null,
                    "container": <head>
                      <style
                        data-emotion="css"
                        data-s=""
                      >
                        
                        .css-jlgjno{padding-bottom:36px;width:100%;-webkit-transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;}
                      </style>
                    </head>,
                    "ctr": 1,
                    "insertionPoint": undefined,
                    "isSpeedy": false,
                    "key": "css",
                    "nonce": undefined,
                    "prepend": undefined,
                    "tags": Array [
                      <style
                        data-emotion="css"
                        data-s=""
                      >
                        
                        .css-jlgjno{padding-bottom:36px;width:100%;-webkit-transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;}
                      </style>,
                    ],
                  },
                }
              }
              isStringTag={true}
              serialized={
                Object {
                  "map": undefined,
                  "name": "jlgjno",
                  "next": undefined,
                  "styles": "padding-bottom:36px;width:100%;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;",
                  "toString": [Function],
                }
              }
            />
            <div
              className="css-jlgjno"
            >
              <Switch>
                <Route
                  computedMatch={
                    Object {
                      "isExact": true,
                      "params": Object {},
                      "path": "/test_link",
                      "url": "/test_link",
                    }
                  }
                  key="test_plugin_name"
                  location={
                    Object {
                      "hash": "",
                      "key": "testKey",
                      "pathname": "/test_link",
                      "search": "",
                    }
                  }
                  path={
                    Array [
                      "/test_link",
                    ]
                  }
                >
                  <Connect(WithAuthComponent)
                    id="test_plugin_name"
                  >
                    <WithAuthComponent
                      id="test_plugin_name"
                      invalidToken={[Function]}
                      loading={false}
                      location="/"
                      loggedIn={true}
                      provider={
                        TestAuthProvider {
                          "authUrl": undefined,
                          "autoLogin": undefined,
                          "mnemonic": undefined,
                          "redirectUrl": null,
                          "token": "logged in",
                          "user": null,
                        }
                      }
                      requestPluginRerender={[Function]}
                      userIsAdmin={true}
                    >
                      <div>
                        <PluginPlaceHolder
                          id="test_plugin_name"
                        >
                          <div
                            id="test_plugin_name"
                          >
                            <Preloader
                              fullScreen={false}
                              loading={true}
                            >
                              Mocked Preloader
                            </Preloader>
                          </div>
                        </PluginPlaceHolder>
                      </div>
                    </WithAuthComponent>
                  </Connect(WithAuthComponent)>
                </Route>
              </Switch>
            </div>
          </Styled(div)>
        </Routing>
      </Connect(Routing)>
    </Router>
  </MemoryRouter>
</Provider>
`;

exports[`Routing component renders a route for admin page 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "testKey",
          "pathname": "/admin",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/admin",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/admin",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Routing)>
        <Routing
          dispatch={[Function]}
          drawerOpen={false}
          loading={false}
          location="/"
          maintenance={
            Object {
              "message": "",
              "show": false,
            }
          }
          plugins={Array []}
          userIsAdmin={true}
          userIsloggedIn={true}
        >
          <Styled(div)
            drawerOpen={false}
          >
            <Insertion
              cache={
                Object {
                  "insert": [Function],
                  "inserted": Object {
                    "jlgjno": true,
                  },
                  "key": "css",
                  "nonce": undefined,
                  "registered": Object {},
                  "sheet": StyleSheet {
                    "_alreadyInsertedOrderInsensitiveRule": true,
                    "_insertTag": [Function],
                    "before": null,
                    "container": <head>
                      <style
                        data-emotion="css"
                        data-s=""
                      >
                        
                        .css-jlgjno{padding-bottom:36px;width:100%;-webkit-transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;}
                      </style>
                    </head>,
                    "ctr": 1,
                    "insertionPoint": undefined,
                    "isSpeedy": false,
                    "key": "css",
                    "nonce": undefined,
                    "prepend": undefined,
                    "tags": Array [
                      <style
                        data-emotion="css"
                        data-s=""
                      >
                        
                        .css-jlgjno{padding-bottom:36px;width:100%;-webkit-transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;}
                      </style>,
                    ],
                  },
                }
              }
              isStringTag={true}
              serialized={
                Object {
                  "map": undefined,
                  "name": "jlgjno",
                  "next": undefined,
                  "styles": "padding-bottom:36px;width:100%;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;",
                  "toString": [Function],
                }
              }
            />
            <div
              className="css-jlgjno"
            >
              <Switch>
                <Route
                  computedMatch={
                    Object {
                      "isExact": true,
                      "params": Object {},
                      "path": "/admin",
                      "url": "/admin",
                    }
                  }
                  exact={true}
                  location={
                    Object {
                      "hash": "",
                      "key": "testKey",
                      "pathname": "/admin",
                      "search": "",
                    }
                  }
                  path="/admin"
                  render={[Function]}
                >
                  <Connect(WithAuthComponent)>
                    <WithAuthComponent
                      invalidToken={[Function]}
                      loading={false}
                      location="/"
                      loggedIn={true}
                      provider={
                        TestAuthProvider {
                          "authUrl": undefined,
                          "autoLogin": undefined,
                          "mnemonic": undefined,
                          "redirectUrl": null,
                          "token": "logged in",
                          "user": null,
                        }
                      }
                      requestPluginRerender={[Function]}
                      userIsAdmin={true}
                    >
                      <div>
                        <Component>
                          Mocked AdminPage
                        </Component>
                      </div>
                    </WithAuthComponent>
                  </Connect(WithAuthComponent)>
                </Route>
              </Switch>
            </div>
          </Styled(div)>
        </Routing>
      </Connect(Routing)>
    </Router>
  </MemoryRouter>
</Provider>
`;

exports[`Routing component renders a route for maintenance page when site is under maintenance and user is not admin 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "testKey",
          "pathname": "/test_link",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/test_link",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/test_link",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Routing)>
        <Routing
          dispatch={[Function]}
          drawerOpen={false}
          loading={false}
          location="/"
          maintenance={
            Object {
              "message": "test",
              "show": true,
            }
          }
          plugins={
            Array [
              Object {
                "displayName": "Test Plugin",
                "link": "/test_link",
                "order": 1,
                "plugin": "test_plugin_name",
                "section": "test section",
              },
            ]
          }
          userIsAdmin={false}
          userIsloggedIn={true}
        >
          <Styled(div)
            drawerOpen={false}
          >
            <Insertion
              cache={
                Object {
                  "insert": [Function],
                  "inserted": Object {
                    "jlgjno": true,
                  },
                  "key": "css",
                  "nonce": undefined,
                  "registered": Object {},
                  "sheet": StyleSheet {
                    "_alreadyInsertedOrderInsensitiveRule": true,
                    "_insertTag": [Function],
                    "before": null,
                    "container": <head>
                      <style
                        data-emotion="css"
                        data-s=""
                      >
                        
                        .css-jlgjno{padding-bottom:36px;width:100%;-webkit-transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;}
                      </style>
                    </head>,
                    "ctr": 1,
                    "insertionPoint": undefined,
                    "isSpeedy": false,
                    "key": "css",
                    "nonce": undefined,
                    "prepend": undefined,
                    "tags": Array [
                      <style
                        data-emotion="css"
                        data-s=""
                      >
                        
                        .css-jlgjno{padding-bottom:36px;width:100%;-webkit-transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;}
                      </style>,
                    ],
                  },
                }
              }
              isStringTag={true}
              serialized={
                Object {
                  "map": undefined,
                  "name": "jlgjno",
                  "next": undefined,
                  "styles": "padding-bottom:36px;width:100%;transition:margin 195ms cubic-bezier(0.4, 0, 1, 1) 0ms,width 195ms cubic-bezier(0.4, 0, 1, 1) 0ms;",
                  "toString": [Function],
                }
              }
            />
            <div
              className="css-jlgjno"
            >
              <Switch>
                <Route
                  component={[Function]}
                  computedMatch={
                    Object {
                      "isExact": false,
                      "params": Object {},
                      "path": "/",
                      "url": "/",
                    }
                  }
                  location={
                    Object {
                      "hash": "",
                      "key": "testKey",
                      "pathname": "/test_link",
                      "search": "",
                    }
                  }
                >
                  <Component
                    history={
                      Object {
                        "action": "POP",
                        "block": [Function],
                        "canGo": [Function],
                        "createHref": [Function],
                        "entries": Array [
                          Object {
                            "hash": "",
                            "key": "testKey",
                            "pathname": "/test_link",
                            "search": "",
                          },
                        ],
                        "go": [Function],
                        "goBack": [Function],
                        "goForward": [Function],
                        "index": 0,
                        "length": 1,
                        "listen": [Function],
                        "location": Object {
                          "hash": "",
                          "key": "testKey",
                          "pathname": "/test_link",
                          "search": "",
                        },
                        "push": [Function],
                        "replace": [Function],
                      }
                    }
                    location={
                      Object {
                        "hash": "",
                        "key": "testKey",
                        "pathname": "/test_link",
                        "search": "",
                      }
                    }
                    match={
                      Object {
                        "isExact": false,
                        "params": Object {},
                        "path": "/",
                        "url": "/",
                      }
                    }
                  >
                    Mocked MaintenancePage
                  </Component>
                </Route>
              </Switch>
            </div>
          </Styled(div)>
        </Routing>
      </Connect(Routing)>
    </Router>
  </MemoryRouter>
</Provider>
`;

exports[`Routing component renders component with no plugin routes 1`] = `
<ContextProvider
  value={null}
>
  <Routing
    dispatch={[Function]}
    drawerOpen={false}
    loading={true}
    location="/"
    maintenance={
      Object {
        "message": "",
        "show": false,
      }
    }
    plugins={Array []}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    userIsAdmin={false}
    userIsloggedIn={false}
  />
</ContextProvider>
`;

exports[`Routing component renders component with plugins 1`] = `
<ContextProvider
  value={null}
>
  <Routing
    dispatch={[Function]}
    drawerOpen={false}
    loading={true}
    location="/"
    maintenance={
      Object {
        "message": "",
        "show": false,
      }
    }
    plugins={
      Array [
        Object {
          "displayName": "Test Plugin",
          "link": "test link",
          "order": 1,
          "plugin": "test_plugin_name",
          "section": "test section",
        },
        Object {
          "displayName": "Test Plugin Alt link",
          "link": "test link alt",
          "order": 2,
          "plugin": "test_plugin_name",
          "section": "test section",
        },
        Object {
          "displayName": "Test Plugin 2",
          "link": "test link 2",
          "order": 3,
          "plugin": "test_plugin_name_2",
          "section": "test section 2",
        },
        Object {
          "admin": true,
          "displayName": "Test Plugin Admin",
          "link": "test link admin",
          "order": 4,
          "plugin": "test_plugin_name",
          "section": "test section",
        },
        Object {
          "admin": true,
          "displayName": "Test Plugin Admin Alt link",
          "link": "test link admin alt",
          "order": 5,
          "plugin": "test_plugin_name",
          "section": "test section",
        },
      ]
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    userIsAdmin={false}
    userIsloggedIn={false}
  />
</ContextProvider>
`;

exports[`Routing component renders placeholder for a plugin 1`] = `
<div
  id="test_id"
>
  <Preloader
    fullScreen={false}
    loading={true}
  />
</div>
`;
